README.md
# Yoda Flutter plugin

Flutter widget that let you slice any kind of child widget in a matrix of tiles and animate them with different kind of ways. Currently with Explode, Vortex and Flakes.
Works on all platforms.

![Screenshot](https://github.com/alnitak/yoda/blob/master/img/yoda.gif?raw=true "yoda Demo")

## Yoda Widget Properties

##### Yoda widget
* [**yodaEffect**] Animation effect to use (see below).
* [**controller**] Animation controller.
* [**duration**] Animation duration.
* [**startWhenTapped**] Enable tap gesture to start animation.
* [**animParameters**] Parameters used to animate tiles (see below).
* [**child**] Whatever widget to slice.

##### YodaEffect enum
* [**Explosion**]
* [**Vortex**]
* [**Flakes**]

##### AnimParameters properties
* [**yodaBarrier**] Define which edges are barraged (see below).
* [**hTiles**] Number of horizontal tiles to divide the child widget.
* [**vTiles**] Number of vertical tiles to divide the child widget.
* [**fractionalCenter**] Default center where force is given when not tapping on it.
* [**gravity**] Vertical force.
* [**effectPower**] Effect strength.
* [**blurPower**] Blur strength.
* [**randomness**] Random strength.

##### YodaBarrier properties
* [**left**] Block tiles moving over the left edge.
* [**right**] Block tiles moving over the right edge.
* [**top**] Block tiles moving over the top edge.
* [**bottom**] Block tiles moving over the bottom edge.


## How to use

If you need to control the animation status:
```dart
YodaController _yodaControllerExplode;
  
@override
  void initState() {
    super.initState();
    _yodaControllerExplode = YodaController()
    ..addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        _yodaControllerExplode.reset();
      }
    });
  }
```
YodaController accept start() and reset() methods.


```dart
Yoda(
  yodaEffect: YodaEffect.Explosion,
  controller: _yodaControllerExplode,
  duration: Duration(milliseconds: 2500),
  animParameters: AnimParameters(
    yodaBarrier: YodaBarrier(bottom: true, left: true, right: true),
    fractionalCenter: Offset(0.5, 1.0),
    hTiles: 20,
    vTiles: 20,
    effectPower: 0.2,
    blurPower: 5,
    gravity: 0.1,
    randomness: 30,
  ),
  startWhenTapped: true,
  child: SizedBox(
    width: 250,
    height: 180,
    child: Image.asset('assets/dash.png', fit: BoxFit.fill)
  )
)
```

----
##### TODO
- Find a better way to capture the widget
- When changing Yoda widget parameters, 2 hot reloads are needed to make it to work!

CHANGELOG.md
## [1.0.2] - 15 Oct 2023
updated some deps
fix capture widget

## [1.0.1] - 3 Feb 2023
Updated Android compile version

## [1.0.0] - 14 Apr 2021
* **Breaking change**: 'power' property renamed to 'effectPower'
* **Breaking change**: YodaController()..addStatusListener now pass Yoda context back to be able for example to remove an item from a grid (see example)
* Example changed
* Added 'blurPower' property
* Added 'randomness' property

## [0.0.1-dev.2] - 9 Apr 2021

* Fixed some linting problems.

## [0.0.1-dev.1] - 9 Apr 2021

* Initial development release.


test/yoda_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:yoda/yoda.dart';

void main() {
  test('adds one to input values', () {

  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:yoda_example/main.dart';

void main() {
}


lib/src/yoda.dart
/*
 * Copyright (c) 2021 Marco Bavagnoli [marcobavagnolidev@gmail.com]. All rights reserved.
 */

import 'dart:async';
import 'dart:math';
import 'dart:typed_data';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:yoda/src/defines.dart';
import 'package:yoda/src/yoda_explode.dart';
import 'package:yoda/src/yoda_flakes.dart';
import 'package:yoda/src/yoda_vortex.dart';

typedef YodaStatusListener = void Function(
    AnimationStatus status, BuildContext context);

class YodaController {
  _YodaState? _yodaState;
  YodaStatusListener? _yodaStatusListener;

  void _addState(_YodaState yodaState) {
    this._yodaState = yodaState;
  }

  /// Determine if the CustomWidgetController is attached to an instance
  /// of the CustomWidget (this property must return true before any other
  /// functions can be used)
  bool get isAttached => _yodaState != null;

  void addStatusListener(YodaStatusListener listener) {
    _yodaStatusListener = listener;
  }

  /// exposed methods
  void start({Offset? center}) {
    assert(isAttached, "YodaController must be attached to a Yoda widget");
    _yodaState?.start(center: center);
  }

  void reset() {
    assert(isAttached, "YodaController must be attached to a Yoda widget");
    _yodaState?.reset();
  }

  Key? getKey() {
    assert(isAttached, "YodaController must be attached to a Yoda widget");
    return _yodaState?.getKey();
  }
}

class Yoda extends StatefulWidget {
  // Effect to use
  final YodaEffect yodaEffect;

  // Animation controller
  final YodaController? controller;

  // Duration of the animation
  final Duration duration;

  // widget to slice
  final Widget child;

  // Animation parameters
  final AnimParameters? animParameters;

  // Enable the animation by tapping the child widget and use
  // the tapping position as the center of the force
  final bool startWhenTapped;

  Yoda({
    required this.yodaEffect,
    this.controller,
    this.duration = const Duration(milliseconds: 1000),
    required this.child,
    this.animParameters,
    this.startWhenTapped = false,
  }) : super(key: UniqueKey()) {
    assert(
        !(yodaEffect == YodaEffect.Flakes &&
            animParameters != null &&
            animParameters!.gravity <= 0),
        "Please, provide a gravity > 0 with Flocks effect");
    assert(animParameters != null && animParameters!.blurPower >= 0,
        "blurPower must be >= 0");
  }

  @override
  _YodaState createState() => _YodaState(controller);
}

class _YodaState extends State<Yoda> with TickerProviderStateMixin {
  YodaController? yodaController;
  late AnimationController controller;
  // late void Function(AnimationStatus) statusListener;
  // late void Function() listener;
  late GlobalKey gk;
  final Completer<bool> completer = Completer<bool>();
  int retryCount = 0;
  CapturedWidged captured = CapturedWidged();
  AnimObject animObject = AnimObject();
  final int rgba32HeaderSize = 122;

  _YodaState(this.yodaController) {
    if (yodaController != null) yodaController?._addState(this);
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();

    gk = GlobalKey();

    init();
  }

  init() {
    controller = AnimationController(vsync: this, duration: widget.duration)
      ..addStatusListener(statusListener)
      ..addListener(listener);

    if (widget.animParameters != null)
      animObject.animParameters = widget.animParameters!;
  }

  void listener() {
    setState(() {});
  }

  void statusListener(AnimationStatus status) {
    if (yodaController != null && yodaController!._yodaStatusListener != null) {
      yodaController!._yodaStatusListener!(status, context);
    }
  }

  // TODO: When changing Yoda widget parameters, 2 hot reloads are needed to make it to work!
  //hot reload?
  @override
  void reassemble() {
    super.reassemble();
    controller.reset();
    controller.removeStatusListener(statusListener);
    controller.removeListener(listener);
    controller.dispose();
    init();
  }

  @override
  Widget build(BuildContext context) {
    CustomPainter? painter;
    if (widget.yodaEffect == YodaEffect.Explosion)
      painter = YodaExplode(
        animObject: animObject,
        controllerValue: controller.value,
      );
    else if (widget.yodaEffect == YodaEffect.Vortex)
      painter = YodaVortex(
        animObject: animObject,
        controllerValue: controller.value,
      );
    else if (widget.yodaEffect == YodaEffect.Flakes)
      painter = YodaFlakes(
        animObject: animObject,
        controllerValue: controller.value,
      );

    return GestureDetector(
      onTapDown: !widget.startWhenTapped
          ? null
          : (TapDownDetails details) {
              start(center: details.localPosition);
            },
      child: RepaintBoundary(
        key: gk,
        child: Stack(
          children: [
            Opacity(
              opacity: controller.value > 0 && captured.size != null ? 0.0 : 1,
              child: widget.child,
            ),
            if (controller.value > 0 && captured.size != null)
              CustomPaint(
                size: captured.size!,
                painter: painter,
              ),
          ],
        ),
      ),
    );
  }

  void start({Offset? center}) {
    if (gk.currentContext == null) return;
    if (controller.isAnimating) {
      controller.reset();
      controller.forward(from: 0);
      return;
    }

    if (center == null) {
      animObject.center = Offset(
          (captured.size?.width ?? 0) *
              animObject.animParameters.fractionalCenter.dx,
          (captured.size?.height ?? 0) *
              animObject.animParameters.fractionalCenter.dy);
    } else {
      animObject.center = center;
    }

    startCapture().then((value) {
      if (value) {
        calcObjectParams();
        controller.forward(from: 0);
      }
    });
  }

  Future<bool> startCapture() async {
    retryCount = 0;
    captured.byteData = null;
    captured.size = null;
    var success = await captureWidget(gk);
    if (success && widget.animParameters != null) {
      await makeTiles();
    }
    return success;
  }

  void reset() {
    controller.reset();
  }

  Key? getKey() {
    return widget.key;
  }

  // TODO: find a better way to capture the widget
  Future<bool> captureWidget(GlobalKey widgetKey) async {
    ui.Image? image;

    try {
      RenderRepaintBoundary? boundary = widgetKey.currentContext
          ?.findRenderObject() as RenderRepaintBoundary?;

      if (retryCount > 30) {
        // how many times to try? 150ms max (15*10ms)
        completer.complete(false);
      }

      if (boundary != null) {
        image = await boundary.toImage();

        captured.byteData =
            await image.toByteData(format: ui.ImageByteFormat.rawRgba);
        captured.size = Size(image.width.toDouble(), image.height.toDouble());
      }

      if (image != null) {
        completer.complete(true);
      } else {
        // with a child like:
        //   SizedBox(
        //       width: 250,
        //       height: 200,
        //       child: Image.asset('assets/dash.png', fit: BoxFit.fill)
        //   )
        // the first 2 boundary acquired are empty
        Timer(Duration(milliseconds: 20), () {
          captureWidget(widgetKey);
        });
        retryCount++;
      }
    } catch (exception) {
      retryCount++;
      //Delay is required. See Issue https://github.com/flutter/flutter/issues/22308
      Timer(Duration(milliseconds: 20), () {
        captureWidget(widgetKey);
      });
    }
    return completer.future;
  }

  // calculate object parameters based on the center of each tiles
  // and the center of the force
  void calcObjectParams() {
    double distance = 0;
    animObject.maxDistance = 0;
    animObject.angle.clear();
    animObject.distance.clear();
    animObject.velocity.clear();
    for (int i = 0; i < animObject.tileUiImages.length; i++) {
      animObject.angle.add(atan2(animObject.offset[i].dy - animObject.center.dy,
          animObject.offset[i].dx - animObject.center.dx));

      distance = sqrt(pow(animObject.offset[i].dx - animObject.center.dx, 2) +
          pow(animObject.offset[i].dy - animObject.center.dy, 2));
      animObject.distance.add(distance);

      if (distance > animObject.maxDistance) animObject.maxDistance = distance;
    }

    // based on maxDistance, calculate the velocity.
    // 0..1   0 at max distance, 1 near the center
    for (int i = 0; i < animObject.tileUiImages.length; i++) {
      animObject.velocity.add(animObject.animParameters.effectPower /
          (animObject.distance[i] / animObject.maxDistance));
    }
  }

  Future<void> makeTiles() async {
    animObject.offset.clear();
    animObject.size.clear();
    animObject.distance.clear();
    animObject.angle.clear();
    animObject.velocity.clear();
    animObject.tileUiImages.clear();
    int xStep = (captured.size?.width ?? 0) ~/ animObject.animParameters.hTiles;
    int yStep =
        (captured.size?.height ?? 0) ~/ animObject.animParameters.vTiles;

    int y1 = 0;
    int y2 = yStep;
    while (y1 <= (captured.size?.height ?? 0) - yStep) {
      // TODO
      // if (y1+yStep>captured.size.height)
      //   y2 = captured.size.height.toInt();
      // else
      y2 = y1 + yStep;

      int x1 = 0;
      int x2 = xStep;
      while (x1 <= (captured.size?.width ?? 0) - xStep) {
        // TODO
        // if (x1+xStep>captured.size.height)
        //   x2 = captured.size.height.toInt();
        // else
        x2 = x1 + xStep;
        await makeTile(x1, y1, x2, y2);
        x1 += xStep;
      }

      y1 += yStep;
    }
  }

  Future<void> makeTile(int x1, int y1, int x2, int y2) async {
    int bytes = 4;
    Uint8List header = bmpHeader(x2 - x1, y2 - y1);
    Uint8List tile =
        Uint8List(rgba32HeaderSize + (x2 - x1) * bytes * (y2 - y1) * bytes);
    ByteData data =
        ByteData(rgba32HeaderSize + (x2 - x1) * bytes * (y2 - y1) * bytes);
    int i = 0;
    for (i = 0; i < rgba32HeaderSize; i++) {
      tile[i] = header[i];
      data.setInt8(i, header[i]);
    }
    int rowBytes = (captured.size?.width ?? 0).toInt() * bytes;
    for (int y = y1; y < y2; y++) {
      for (int x = x1 * bytes; x < x2 * bytes; x++) {
        tile[i] = captured.byteData!.getInt8(y * rowBytes + x);
        data.setInt8(i, captured.byteData!.getInt8(y * rowBytes + x));
        i++;
      }
    }

    // copy from decodeImageFromList of package:flutter/painting.dart
    final ui.Codec codec =
        await ui.instantiateImageCodec(Uint8List.fromList(tile));
    final ui.FrameInfo frameInfo = await codec.getNextFrame();
    animObject.tileUiImages.add(frameInfo.image);

    Size size = Size(x2.toDouble() - x1, y2.toDouble() - y1);
    Offset offset =
        Offset(x1.toDouble() + size.width / 2, y1.toDouble() + size.height / 2);
    animObject.size.add(size);
    animObject.offset.add(offset);

    // if not tappable, calculate now the distance from center. Else calculate it at tapping time
    if (!widget.startWhenTapped) {
      animObject.angle.add(atan2(
          offset.dy - animObject.center.dy, offset.dx - animObject.center.dx));

      animObject.distance.add(sqrt(pow(offset.dx - animObject.center.dx, 2) +
          pow(offset.dy - animObject.center.dy, 2)));
    }
  }

  Uint8List bmpHeader(int width, int height) {
    int contentSize = width * height;
    Uint8List ret = Uint8List(rgba32HeaderSize);

    final ByteData bd = ret.buffer.asByteData();
    bd.setUint8(0x0, 0x42);
    bd.setUint8(0x1, 0x4d);
    bd.setInt32(0x2, contentSize + rgba32HeaderSize, Endian.little);
    bd.setInt32(0xa, rgba32HeaderSize, Endian.little);
    bd.setUint32(0xe, 108, Endian.little);
    bd.setUint32(0x12, width, Endian.little);
    bd.setUint32(0x16, -height, Endian.little);
    bd.setUint16(0x1a, 1, Endian.little);
    bd.setUint32(0x1c, 32, Endian.little); // pixel size
    bd.setUint32(0x1e, 3, Endian.little); //BI_BITFIELDS
    bd.setUint32(0x22, contentSize, Endian.little);
    bd.setUint32(0x36, 0x000000ff, Endian.little);
    bd.setUint32(0x3a, 0x0000ff00, Endian.little);
    bd.setUint32(0x3e, 0x00ff0000, Endian.little);
    bd.setUint32(0x42, 0xff000000, Endian.little);

    return bd.buffer.asUint8List();
  }
}


lib/src/defines.dart
/*
 * Copyright (c) 2021 Marco Bavagnoli [marcobavagnolidev@gmail.com]. All rights reserved.
 */

import 'dart:typed_data';
import 'dart:ui' as ui;

import 'package:flutter/painting.dart';

enum YodaEffect {
  Explosion,
  Vortex,
  Flakes,
}

/// Tiles barriers
class YodaBarrier {
  // block tiles moving over the left edge
  final bool left;

  // block tiles moving over the right edge
  final bool right;

  // block tiles moving over the top edge
  final bool top;

  // block tiles moving over the bottom edge
  final bool bottom;

  const YodaBarrier({
    this.left = false,
    this.right = false,
    this.top = false,
    this.bottom = false,
  });
}

class AnimParameters {
  // define which edges are barraged
  final YodaBarrier yodaBarrier;

  // number of horizontal tiles to divide the child widget
  final int hTiles;

  // number of vertical tiles to divide the child widget
  final int vTiles;

  // default center where force is given
  Offset fractionalCenter;

  // vertical force
  double gravity;

  // strength of the effect
  double effectPower;

  // strength of the blur
  double blurPower;

  // shifting randomness
  int randomness;

  AnimParameters(
      {this.yodaBarrier = const YodaBarrier(),
      this.hTiles = 0,
      this.vTiles = 0,
      this.fractionalCenter = const Offset(0.5, 0.5),
      this.gravity = 9.81,
      this.effectPower = 1.0,
      this.blurPower = 0.0,
      this.randomness = 0});
}

class AnimObject {
  // Animation parameters (see above)
  AnimParameters animParameters = AnimParameters();

  // Position where the force act inside the widget
  Offset center = Offset.zero;

  // Calculated max distance of a tile from the center
  double maxDistance = 0;

  // Calculated offsets of each tiles
  List<Offset> offset = [];

  // Calculated sizes of each tiles
  List<Size> size = [];

  // Calculated distance of each tiles
  List<double> distance = [];

  // Calculated angles of each tiles relative to center
  List<double> angle = [];

  // Calculated velocity of each tiles based on the nearness from the center
  // 0 is when the tile is at the maxDistance, 1 when it is in the center
  List<double> velocity = [];

  // Tiles images
  List<ui.Image> tileUiImages = [];
}

class CapturedWidged {
  ByteData? byteData; // uncompressed 32bit RGBA image data
  Size? size;
}


lib/src/yoda_flakes.dart
/*
 * Copyright (c) 2021 Marco Bavagnoli [marcobavagnolidev@gmail.com]. All rights reserved.
 */

import 'dart:math';

import 'package:flutter/animation.dart';
import 'package:flutter/rendering.dart';
import 'package:yoda/src/defines.dart';

class YodaFlakes extends CustomPainter {
  final AnimObject animObject;
  final double controllerValue;

  YodaFlakes({
    required this.animObject,
    required this.controllerValue,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final Paint paint = Paint();
    Cubic opacityCurve = Curves.easeInOutCubic;
    Curve velocityCurve = Curves.easeInQuad;

    for (int y = 0; y < animObject.animParameters.vTiles; y++) {
      for (int x = 0; x < animObject.animParameters.hTiles; x++) {
        int tileID = y * animObject.animParameters.vTiles + x;

        double newX = animObject.offset[tileID].dx +
            sin(velocityCurve.transform(1 - controllerValue) *
                    animObject.animParameters.effectPower) *
                y;

        double newY = animObject.offset[tileID].dy +
            y *
                velocityCurve.transform(controllerValue) *
                Random(tileID).nextDouble() *
                animObject.animParameters.randomness;

        if (animObject.animParameters.yodaBarrier.right && newX > size.width)
          newX = size.width;
        if (animObject.animParameters.yodaBarrier.left && newX < 0) newX = 0;
        if (animObject.animParameters.yodaBarrier.bottom && newY > size.height)
          newY = size.height;
        if (animObject.animParameters.yodaBarrier.top && newY < 0) newY = 0;

        Offset offset = Offset(newX, newY);

        // opacity
        paint.color = Color.fromRGBO(
            0, 0, 0, 1 - opacityCurve.transformInternal(controllerValue));

        // blur
        if (animObject.animParameters.blurPower > 0)
          paint.maskFilter = MaskFilter.blur(
              BlurStyle.normal,
              (velocityCurve.transform(controllerValue)) *
                  animObject.animParameters.blurPower);

        canvas.drawImage(
            animObject.tileUiImages[tileID],
            offset -
                Offset(animObject.size[tileID].width / 2,
                    animObject.size[tileID].height / 2),
            paint);
      }
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}


lib/src/yoda_vortex.dart
/*
 * Copyright (c) 2021 Marco Bavagnoli [marcobavagnolidev@gmail.com]. All rights reserved.
 */

import 'dart:math';

import 'package:flutter/animation.dart';
import 'package:flutter/rendering.dart';
import 'package:yoda/src/defines.dart';

class YodaVortex extends CustomPainter {
  final AnimObject animObject;
  final double controllerValue;

  YodaVortex({
    required this.animObject,
    required this.controllerValue,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final Paint paint = Paint();
    Cubic opacityCurve = Curves.easeInOutCubic;
    Curve velocityCurve = Curves.decelerate;

    for (int i = 0; i < animObject.tileUiImages.length; i++) {
      double newDistance =
          -animObject.distance[i] * (velocityCurve.transform(controllerValue));

      // randomize the new distance
      if (animObject.animParameters.randomness > 0)
        newDistance +=
            (Random(i).nextDouble() * animObject.animParameters.randomness);

      double dy = (controllerValue) *
          animObject.distance[i] *
          animObject.animParameters.gravity;
      double newAngle = animObject.angle[i] +
          (newDistance /
              animObject.maxDistance *
              animObject.animParameters.effectPower);

      double newX = animObject.offset[i].dx + cos(newAngle) * newDistance;
      double newY = animObject.offset[i].dy + sin(newAngle) * newDistance + dy;

      if (animObject.animParameters.yodaBarrier.right && newX > size.width)
        newX = size.width;
      if (animObject.animParameters.yodaBarrier.left && newX < 0) newX = 0;
      if (animObject.animParameters.yodaBarrier.bottom && newY > size.height)
        newY = size.height;
      if (animObject.animParameters.yodaBarrier.top && newY < 0) newY = 0;

      Offset offset = Offset(newX, newY);

      // opacity
      paint.color =
          Color.fromRGBO(0, 0, 0, 1 - opacityCurve.transform(controllerValue));

      // blur
      if (animObject.animParameters.blurPower > 0)
        paint.maskFilter = MaskFilter.blur(
            BlurStyle.normal,
            (velocityCurve.transform(controllerValue)) *
                animObject.animParameters.blurPower);

      canvas.drawImage(
          animObject.tileUiImages[i],
          offset -
              Offset(
                  animObject.size[i].width / 2, animObject.size[i].height / 2),
          paint);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}


lib/src/yoda_explode.dart
/*
 * Copyright (c) 2021 Marco Bavagnoli [marcobavagnolidev@gmail.com]. All rights reserved.
 */

import 'dart:math';

import 'package:flutter/animation.dart';
import 'package:flutter/rendering.dart';
import 'package:yoda/src/defines.dart';

class YodaExplode extends CustomPainter {
  final AnimObject animObject;
  final double controllerValue;

  YodaExplode({
    required this.animObject,
    required this.controllerValue,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final Paint paint = Paint();
    Cubic opacityCurve = Curves.easeInOutCubic;
    Cubic velocityCurve = Curves.easeOutCirc;

    for (int i = 0; i < animObject.tileUiImages.length; i++) {
      double newDistance = animObject.distance[i] *
          (animObject.velocity[i] * velocityCurve.transform(controllerValue));

      // randomize the new distance
      if (animObject.animParameters.randomness > 0)
        newDistance +=
            (Random(i).nextDouble() * animObject.animParameters.randomness);

      double dy = (controllerValue) *
          animObject.distance[i] *
          animObject.animParameters.gravity;

      double newX =
          animObject.offset[i].dx + cos(animObject.angle[i]) * newDistance;
      double newY =
          animObject.offset[i].dy + sin(animObject.angle[i]) * newDistance + dy;

      if (animObject.animParameters.yodaBarrier.right && newX > size.width)
        newX = size.width;
      if (animObject.animParameters.yodaBarrier.left && newX < 0) newX = 0;
      if (animObject.animParameters.yodaBarrier.bottom && newY > size.height)
        newY = size.height;
      if (animObject.animParameters.yodaBarrier.top && newY < 0) newY = 0;

      Offset offset = Offset(newX, newY);

      // opacity
      paint.color =
          Color.fromRGBO(0, 0, 0, 1 - opacityCurve.transform(controllerValue));

      // blur
      if (animObject.animParameters.blurPower > 0)
        paint.maskFilter = MaskFilter.blur(
            BlurStyle.normal,
            (velocityCurve.transform(controllerValue)) *
                animObject.animParameters.blurPower);

      canvas.drawImage(
          animObject.tileUiImages[i],
          offset -
              Offset(
                  animObject.size[i].width / 2, animObject.size[i].height / 2),
          paint);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:yoda/yoda.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: YodaExample(),
    );
  }
}

class YodaExample extends StatefulWidget {
  @override
  _YodaExampleState createState() => _YodaExampleState();
}

class _YodaExampleState extends State<YodaExample> {
  List<Yoda> items = [];

  @override
  void initState() {
    super.initState();
    buildItems();
  }

  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;

    return Scaffold(
      appBar: AppBar(
        title: Text('Yoda example'),
      ),
      body: SingleChildScrollView(
        child: Wrap(
          alignment: WrapAlignment.center,
          children: [
            for (final item in items)
              SizedBox(width: width / 2, height: width / 2 * 0.9, child: item)
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          setState(() {
            buildItems();
          });
        },
        child: Icon(Icons.add),
      ),
    );
  }

  buildItems() {
    items.add(
      Yoda(
        yodaEffect: YodaEffect.Explosion,
        controller: YodaController()
          ..addStatusListener((status, context) {
            if (status == AnimationStatus.completed) {
              items.removeWhere((element) {
                bool isEqual =
                    element.controller!.getKey() == context.widget.key;
                if (isEqual) setState(() {});
                return isEqual;
              });
            }
          }),
        duration: Duration(milliseconds: 2000),
        animParameters: AnimParameters(
            yodaBarrier: YodaBarrier(bottom: true),
            fractionalCenter: Offset(0.5, 1.0),
            hTiles: 20,
            vTiles: 20,
            effectPower: 0.2,
            gravity: 0.1,
            blurPower: 5,
            randomness: 30),
        startWhenTapped: true,
        child: YodaCard(assetName: 'assets/dash.png'),
      ),
    );

    items.add(
      Yoda(
          yodaEffect: YodaEffect.Vortex,
          controller: YodaController()
            ..addStatusListener((status, context) {
              if (status == AnimationStatus.completed) {
                items.removeWhere((element) {
                  bool isEqual =
                      element.controller!.getKey() == context.widget.key;
                  if (isEqual) setState(() {});
                  return isEqual;
                });
              }
            }),
          duration: Duration(milliseconds: 2500),
          animParameters: AnimParameters(
              yodaBarrier:
                  YodaBarrier(top: true, bottom: true, left: true, right: true),
              fractionalCenter: Offset(0.5, 1.0),
              hTiles: 20,
              vTiles: 20,
              effectPower: 10,
              gravity: 0,
              blurPower: 5,
              randomness: 20),
          startWhenTapped: true,
          child: YodaCard(assetName: 'assets/dash2.png')),
    );

    items.add(
      Yoda(
          yodaEffect: YodaEffect.Flakes,
          controller: YodaController()
            ..addStatusListener((status, context) {
              if (status == AnimationStatus.completed) {
                items.removeWhere((element) {
                  bool isEqual =
                      element.controller!.getKey() == context.widget.key;
                  if (isEqual) setState(() {});
                  return isEqual;
                });
              }
            }),
          duration: Duration(milliseconds: 2500),
          animParameters: AnimParameters(
              yodaBarrier: YodaBarrier(),
              // fractionalCenter: Offset(0.5, 1.0), // not used in the Flocks effect
              hTiles: 20,
              vTiles: 20,
              effectPower: 10,
              gravity: 30,
              blurPower: 30,
              randomness: 30),
          startWhenTapped: true,
          child: YodaCard(
            assetName: 'assets/dash3.png',
          )),
    );

    items.add(
      Yoda(
        yodaEffect: YodaEffect.Explosion,
        controller: YodaController()
          ..addStatusListener((status, context) {
            if (status == AnimationStatus.completed) {
              items.indexWhere((element) {
                bool isEqual =
                    element.controller!.getKey() == context.widget.key;
                if (isEqual) element.controller!.reset();
                return isEqual;
              });
            }
          }),
        duration: Duration(milliseconds: 1000),
        animParameters: AnimParameters(
            yodaBarrier: YodaBarrier(bottom: true),
            fractionalCenter: Offset(0.5, 1.0),
            hTiles: 20,
            vTiles: 20,
            effectPower: 0.1,
            gravity: 0.5,
            blurPower: 0,
            randomness: 0),
        startWhenTapped: true,
        child: YodaCard(assetName: 'assets/dash.png'),
      ),
    );

    items.add(
      Yoda(
        yodaEffect: YodaEffect.Flakes,
        controller: YodaController()
          ..addStatusListener((status, context) {
            if (status == AnimationStatus.completed) {
              items.indexWhere((element) {
                bool isEqual =
                    element.controller!.getKey() == context.widget.key;
                if (isEqual) element.controller!.reset();
                return isEqual;
              });
            }
          }),
        duration: Duration(milliseconds: 1000),
        animParameters: AnimParameters(
            yodaBarrier: YodaBarrier(),
            fractionalCenter: Offset(0.5, 1.0),
            hTiles: 20,
            vTiles: 20,
            effectPower: 0,
            gravity: 2,
            blurPower: 0,
            randomness: 80),
        startWhenTapped: true,
        child: YodaCard(assetName: 'assets/dash.png'),
      ),
    );

    items.add(
      Yoda(
          yodaEffect: YodaEffect.Vortex,
          controller: YodaController()
            ..addStatusListener((status, context) {
              if (status == AnimationStatus.completed) {
                items.indexWhere((element) {
                  bool isEqual =
                      element.controller!.getKey() == context.widget.key;
                  if (isEqual) element.controller!.reset();
                  return isEqual;
                });
              }
            }),
          duration: Duration(milliseconds: 2500),
          animParameters: AnimParameters(
              yodaBarrier:
                  YodaBarrier(top: true, bottom: true, left: true, right: true),
              fractionalCenter: Offset(0.5, 1.0),
              hTiles: 30,
              vTiles: 30,
              effectPower: 5,
              gravity: 0,
              blurPower: 0,
              randomness: 10),
          startWhenTapped: true,
          child: YodaCard(assetName: 'assets/dash.png')),
    );
  }
}

class YodaCard extends StatelessWidget {
  final String assetName;

  YodaCard({
    required this.assetName,
  });

  @override
  Widget build(BuildContext context) {
    return Card(
      elevation: 10,
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(15.0),
      ),
      child: Column(
        children: [
          Text(
            'Flutter is awesome',
            style: TextStyle(fontWeight: FontWeight.bold),
            textScaleFactor: 1.2,
          ),
          Image.asset(assetName, fit: BoxFit.fitHeight),
        ],
      ),
    );
  }
}
